Feat(Jest-runner): New implementation for jest-runner

Archcry opened the pull request:
This is a new implementation for the jest runner that also support custom configurations and react projects. It also uses a higher implementation level of Jest so we no longer need the haste context.

Resolves https://github.com/stryker-mutator/stryker-jest-runner/issues/24

Archcry commented:
It seems like Bithound does not really understand how interfaces in typescript work, the interfaces are correct when running the tslint.json from the master branch.

mthmulders commented on src/configLoaders/DefaultConfigLoader.ts:
```diff
@@ -0,0 +1,20 @@
+    return JSON.stringify(this.getJestConfigFromPackageJson(this.projectRoot));
+  }
+
+  private getJestConfigFromPackageJson(projectRoot: string) {
```
Does this work with the [Multi-Project-Runner & Configuration Overhaul](http://facebook.github.io/jest/blog/2017/05/06/jest-20-delightful-testing-multi-project-runner.html) in Jest 20?

Archcry replied:
Well I did see an option to add multiple projectroots (as an array) to the run function of Jest. Currently I programmed it so that only one projectroot is provided in that array. Is there a case where we want to test multiple projects at the same time with Stryker?

mthmulders replied:
I don't really think we need it immediately, as long as we have an option to add / extend it later on.

Archcry replied:
That should be possible, we could add a `projectroot` property in the future which will accept a string or array of strings. I am not sure on how this will work with the jest configuration in the `package.json` though. But since the configuration from the jest config is loaded this shouldn't be a problem™

mthmulders commented on src/configLoaders/ReactConfigLoader.ts:
```diff
@@ -0,0 +1,31 @@
+import createReactJestConfig from '../utils/createReactJestConfig';
+import * as path from 'path';
+
+export default class ReactConfigLoader implements ConfigLoader {
```
Can you add some documentation: what does this class do, and why does it do so? I can't figure out why it exists.

Archcry replied:
The `ConfigEditor` interacts with an implementation of `ConfigLoader` to load the configuration for a certain type of project specified in the `stryker.conf.js` ([see readme#configuration](https://github.com/stryker-mutator/stryker-jest-runner/blob/feat-new-jest-runner/README.MD#configuration)).

Personally I am not a big fan of writing comments in code because comments almost always lie. What would be a good place to document this?

mthmulders replied:
Personally, I prefer source code comments over people having to reverse-engineer it. Having them next to the code makes it easier to update it as soon as the code changes, while having it in a separate file makes that a bit harder.

mthmulders commented on src/utils/createReactJestConfig.ts:
```diff
@@ -0,0 +1,5 @@
+export default function createReactJestConfig(resolve: Function, projectRoot: string, ejected: boolean, loader?: NodeRequire): string {
```
Do I understand correctly that this function is specifically targeted at people who use [Create React App](https://github.com/facebookincubator/create-react-app)? How about also supporting users of its [TypeScript counterpart](https://github.com/wmonk/create-react-app-typescript)?

Archcry replied:
Yes it is targetted towards the users that use `create-react-app` to create their react apps. Adding support for the typescript counterpart sounds good. I'd have to investigate this though. It might be better to merge this and make an issue for typescript support as I am not sure I can add this in the last two weeks of my graduation.

Archcry replied:
People can always eject their project to get a jest config via the `package.json`, I realise this isn't ideal because you'd lose all functionality that is provided by `react-scripts` but it's a way to mutation-test the project.

mthmulders commented on src/configLoaders/ConfigLoader.ts:
```diff
@@ -0,0 +1,3 @@
+export default interface ConfigLoader {
```
Can you explain what a `ConfigLoader` is and why it is needed? Maybe add some references to Jest documentation or code?

Archcry replied:
Well, this is not necessarily Jest related. In the code we use an implementation of `ConfigLoader` to return either a react config (generated via the `ReactConfigLoader`) or the config in the `package.json` (via the `DefaultConfigLoader`). We can add more config loaders for different kinds of projects in the future when the configuration is not present in the package.json.

Archcry replied:
We can also add a `ConfigLoader` implementation for the Typescript flavour of react in the future. I made it this way so that we can add new loaders on the fly.

mthmulders replied:
That explanation makes sense. I'd like to see it in source code comments, so people don't have to refer to this PR to find out why the interface exists.

Archcry replied:
That's fine, I'll add the comments this evening hopefully.

mthmulders commented:
I'm pretty impressed by your work. It looks good 👍 . The overall approach is nice, and provides us a looser coupling to React internals, which is a great thing to see!

Would you mind adding some more documentation? For example, I would like the `ReactConfigLoader` and the `DefaultConfigEditor` explaining why they exist and what they do. That might make it easier for other to contribute even more config editors in the future.

Archcry commented:
I did explain that in the last two commits. Did you see them? Or is this just a subtle hint that more documentation is required? :P

simondel commented on package.json:
```diff
@@ -1,78 +1,72 @@
+  "private": true,
   "name": "stryker-jest-runner",
-  "version": "0.3.0",
+  "version": "0.0.0",
```
Let's set this to to 0.3.0?

simondel commented on package.json:
```diff
@@ -1,78 +1,72 @@
+    "@types/sinon": "^2.3.6",
+    "chai": "^4.1.2",
+    "jest": "^20.0.4",
+    "mocha": "^3.5.3",
```
Can you run against mocha 4? (and upgrade the types)

Archcry replied:
Should work, will test.

Archcry replied:
Works with mocha 5, updated the dependency in package.json accordingly.

simondel commented on package.json:
```diff
@@ -1,78 +1,72 @@
-    "Philipp Weissenbacher <philipp.weissenbacher@gmail.com>"
-  ]
+  "dependencies": {
+    "semver": "^5.4.1"
```
Don't you log anything? (You removed the log4js dependency)

Archcry replied:
I currently do not log anything

simondel commented on package.json:
```diff
@@ -1,78 +1,72 @@
 {
+  "private": true,
```
You can remove this

simondel commented on README.MD:
````diff
@@ -0,0 +1,62 @@
+Set your test runner to `jest` in the stryker config, optionally provide the project you're working on for a preset configuration.
+We currently only support `react` preset, enable it by adding the following config entry to the stryker config:
+
+```javascript
````
Could you add two bits of config? One for react projects (which also sets the testRunner property) and one for any other project?

simondel replied:
You specify it later on, but I would expect the testRunner property here as well.

simondel commented on README.MD:
````diff
@@ -0,0 +1,62 @@
+Note that coverageAnalysis is not yet supported, so you must explicitly set it to off in your Stryker configration. Pull requests are - obviously - welcome.
+
+```javascript
+module.exports = function (config) {
````
As mentioned, before, you could combine these fragments. Should you also disable coverageAnalylsis if users choose jest with the stryker cli quickstart?

simondel replied:
Also, it may be smart to specify that users should not set the testRunner property, right @nicojs?

simondel commented on README.MD:
````diff
@@ -0,0 +1,62 @@
+```
+
+## Contributing
+Before you start hacking along, make sure to install a supported version of Jest inside your working copy. The versions supported are listed above. Also make sure to read the Stryker contribution guidelines located in the [Stryker mono repository](https://github.com/stryker-mutator/stryker/blob/master/CONTRIBUTING.md).
````
The default steps (`npm i`, `npm run build`, `npm test`) should be enough to get it up and running, so if that's the case you can remove the note about installing jest.

simondel commented on src/JestConfigEditor.ts:
```diff
@@ -0,0 +1,35 @@
+      project = config.jest.project;
+    }
+
+    config.jest = config.jest || {};
```
If you move this line to the start of the method, you can remove the if-statement above and simply write `let project = config.jest.project || 'default';` if I'm correct.

Archcry replied:
Well if the jest property does not exist it will return with an error saying that it could not find `project` on `undefined`.

Archcry replied:
This would work though:

```javascript
  public edit(strykerConfig: Config): void {
    strykerConfig.jest = strykerConfig.jest || {};
    strykerConfig.jest.project = strykerConfig.jest.project || 'default';
    strykerConfig.jest.config = strykerConfig.jest.config || this.getConfigLoader(strykerConfig.jest.project).loadConfig();
  }
```

simondel commented on src/JestConfigEditor.ts:
```diff
@@ -0,0 +1,35 @@
+import ReactConfigLoader from './configLoaders/ReactConfigLoader';
+
+export default class JestConfigEditor implements ConfigEditor {
+  public edit(config: Config): void {
```
Perhaps it would be clearer to rename `config` to `strykerConfig`. That way you wont have `config.jest.config` and it will be clear what each config really is about.

simondel commented on src/JestConfigEditor.ts:
```diff
@@ -0,0 +1,35 @@
+    let configLoader: ConfigLoader;
+
+    switch (project.toLowerCase()) {
+      case 'default':
```
You may want to define this once

Archcry replied:
What do you mean?

simondel commented on src/jestTestAdapters/JestTestAdapterFactory.ts:
```diff
@@ -0,0 +1,37 @@
+  public static getJestTestAdapter(loader?: NodeRequire): JestTestAdapter {
+    let jestTestAdapter: JestTestAdapter;
+
+    switch (this.getJestType(loader)) {
```
Do you really need to call that method? Now you're coupling two methods together with magic strings.

Archcry replied:
No I don't need to call the method, should refactor that.

simondel commented on src/jestTestAdapters/JestTestAdapterFactory.ts:
```diff
@@ -0,0 +1,37 @@
+import * as semver from 'semver';
+
+export default class JestTestAdapterFactory {
+  public static getJestTestAdapter(loader?: NodeRequire): JestTestAdapter {
```
This loader stuff is for testing right?

Archcry replied:
Yep

simondel commented on src/jestTestAdapters/JestTestAdapterFactory.ts:
```diff
@@ -0,0 +1,37 @@
+
+    switch (this.getJestType(loader)) {
+      case 'promise':
+        jestTestAdapter = new JestPromiseAdapter;
```
Why no `( )`? 😭

Archcry replied:
I guess it's a matter of taste, but I am open for both notations. What do you want me to change it to?

simondel commented on src/JestTestRunner.ts:
```diff
@@ -1,131 +1,49 @@
-      );
-    });
-    return Promise.all(promises);
+    return this.processRunResults(results);
```
Right now, you call a method that just returns an object. I would prefer just returning the object here

Archcry replied:
So assign the output of the function to an extra variable and return that?

simondel commented on src/configLoaders/DefaultConfigLoader.ts:
```diff
@@ -0,0 +1,20 @@
+  private fs: any;
+  private projectRoot: string;
+
+  constructor(projectRoot: string, fs: any) {
```
Could you set the properties in the constructor using ` constructor(private projectRoot: string, private fs: any) {`

Archcry replied:
I really prefer it the way it is right now, it makes the code much clearer to people that have less experience with TypeScript. I mean, what other programming language assign their private variables in the constructor?

simondel commented on src/configLoaders/DefaultConfigLoader.ts:
```diff
@@ -0,0 +1,20 @@
+    return JSON.stringify(this.getJestConfigFromPackageJson(this.projectRoot));
+  }
+
+  private getJestConfigFromPackageJson(projectRoot: string) {
```
The fact that you need two methods to do this is an indicator that it's too complex.
This line for example: `JSON.parse(this.fs.readFileSync(path.join(projectRoot, 'package.json'), 'utf8')).jest;` I have no clue what the object is that we're calling `jest` on

simondel commented on src/configLoaders/ReactConfigLoader.ts:
```diff
@@ -0,0 +1,35 @@
+    this.projectRoot = projectRoot;
+  }
+
+  public loadConfig(): string {
```
Some comments in this method are not useful to me. For example the comment above `return JSON.stringify(jestConfiguration);`.

simondel commented on src/configLoaders/ReactConfigLoader.ts:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  private createJestConfig(reactScriptsLocation: string): any {
+    return createReactJestConfig(
```
You exported a single `this.loader.resolve` call to a different file, in a different folder]. Why is it not a part of this file?

Archcry replied:
This is because I want to be able to make a stub of the dependency in my unit tests. When I don't do this I cannot stub the method.

simondel commented on stryker.conf.js:
```diff
@@ -0,0 +1,21 @@
+      { pattern: 'src/**/*.ts', included: false, mutated: true },
+      '!./src/index.ts',
+
+      // Stryker crashes on integration tests, don't run them
```
Feel free to remove this comment

Archcry replied:
Will do

Archcry commented on .travis.yml:
```diff
@@ -1,15 +1,13 @@
-  - npm i -D jest@${JEST_VERSION} jest-cli@${JEST_VERSION}
\ No newline at end of file
+before_install:
+- if [[ `npm -v` = 2* ]]; then npm i -g npm@3; fi
```
I don't know, I though this was for when you want to test with node 4 within travis.

simondel commented on test/unit/jestTestAdapters/JestCallbackTestAdapterSpec.ts:
```diff
@@ -0,0 +1,62 @@
+
+  let jestCallbackTestAdapter: JestCallbackTestAdapter;
+
+  let projectRoot = '/path/to/project';
```
You can use const in some places here

simondel commented on test/integration/StrykerJestRunnerSpec.ts:
```diff
@@ -0,0 +1,55 @@
+
+  process.env.BABEL_ENV = 'test';
+
+  beforeEach(() => {
```
Having a beforeEach and afterEach for just one spec seems like overkill.

nicojs replied:
But it makes it easier to add tests later on right? Why not leave it? Also: this file already has  more tests.

Archcry replied:
It had only one test before, I added a test in the last revision.

Archcry commented on .travis.yml:
```diff
@@ -1,15 +1,13 @@
-  - JEST_VERSION=20.0.0
-  - JEST_VERSION=21.0.0
+- node
+- '7'
```
What do you mean?

Archcry commented on test/unit/jestTestAdapters/JestCallbackTestAdapterSpec.ts:
```diff
@@ -0,0 +1,62 @@
+  it('should set reporters to an empty array', async () => {
+    await jestCallbackTestAdapter.run(jestConfig, projectRoot);
+
+    expect(jestConfig.reporters).to.be.an('array').that.is.empty;
```
I'll leave it like this, the way it is now shows intent, comparing to an empty array doesn't.

Archcry commented:
Most of the feedback is processed, the Jest configuration is now placed on the Stryker config as an object instead of a string. I think we only need those integration tests now.

Archcry commented:
I added the sampleProject from the current jest-runner to the project and tested it with an extra integration test. I do not have an end-to-end test with Stryker and I would suggest adding it when we merge the Jest runner into the Stryker mono repo because the `run-integration-test.ts` script is there (https://github.com/stryker-mutator/stryker/blob/master/integrationTest/run-integration-test.ts).

This pull-request should be good to go.

Archcry commented:
I added the sampleProject from the current jest-runner to the project and tested it with an extra integration test. I do not have an end-to-end test with Stryker and I would suggest adding it when we merge the Jest runner into the Stryker mono repo because the `run-integration-test.ts` script is there (https://github.com/stryker-mutator/stryker/blob/master/integrationTest/run-integration-test.ts).

I also added support for runtime errors, it will now correctly report a jest runtime error as a runtime error in Stryker!

This pull-request should be good to go.

nicojs commented on src/JestTestRunner.ts:
```diff
@@ -1,131 +1,45 @@
-      status: RunStatus.Error,
-      errorMessages: [ `${error.name}: ${error.message}` ]
+      tests: this.processTestResults(results.testResults),
+      status: (results.numRuntimeErrorTestSuites > 0) ? RunStatus.Error : RunStatus.Complete
```
Is there a way to get at the actual error messages? How would people start to debug when there is a runtime error?

Archcry replied:
It only returns a failureMessage for the individual test suites, not for the run as a whole.

nicojs replied:
are there no runtime error messages on the run result or test file result? Seems strange to me that they do show the number of runtime errors, but not the errors themselves.

Archcry replied:
They are on the fileResult, but there are more fileResults in a single runResult.

nicojs commented on src/configLoaders/DefaultConfigLoader.ts:
```diff
@@ -0,0 +1,29 @@
+    const packageJson = this._fs.readFileSync(path.join(this._projectRoot, 'package.json'), 'utf8');
+
+    // Parse the package.json and return the Jest property
+    const jestConfig = JSON.parse(packageJson).jest;
```
Does this mean we will only support jest config in 'package.json'? Not a separate "jest.config.js" file like explained here: https://facebook.github.io/jest/docs/en/configuration.html

Seems fairly strait forward to support that as well.

Archcry replied:
Well, you can assign the config object manually in the stryker config
```javascript
jest: {
  config: require('path/to/jestConfig.js')
}
```

nicojs replied:
Good call. Can we add it as an example in the readme?

Archcry replied:
I think we can :+1:

nicojs commented on src/configLoaders/JestConfiguration.ts:
```diff
@@ -0,0 +1,60 @@
+// Grabbed from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts
```
We should move this to definitely typed later.

Archcry replied:
It's already there but it doesn't work when installing it unfortunately... I think the types need to be updated.

nicojs replied:
Well, lets do that than 👍 so we can remove this later on

mthmulders replied:
This remark is still relevant. If the type defs from DT are (still) not up to date, I'd at least expect an issue for Stryker Jest runner saying we should use them as soon as they work (again).

Archcry replied:
https://github.com/stryker-mutator/stryker-jest-runner/issues/29

nicojs commented on test/unit/JestTestRunnerSpec.ts:
```diff
@@ -0,0 +1,108 @@
+  });
+
+  it('should call the jestTestRunner run method and return a correct runResult', async () => {
+    runJestStub.returns(Promise.resolve({ results: fakeResults.createSuccessResult() }));
```
It's better to use `stub.resolves(...)` instead of `stub.returns(Promise.resolve(...))`

nicojs commented:
@Floriangomis is also waiting for this feature. See #28

Archcry commented:
I push failureMessages from the individual test suites to the `errorMessages` array in the `RunResult` now, this should be enough to inform the user. I also updated the Readme.md to include the possible options you can set for the jest-runner.

nicojs commented on README.md:
```diff
@@ -21,19 +21,34 @@ The stryker-jest-runner is a plugin for Stryker to enable Jest as a test runner.
+
+| option | value |
+|----|----|
+| project (optional) | The project you are working on (we currently only support "react")|
```
This isn't true, as we also support "Default" right? Probably we should also explain what "react" does (mimic `react-scripts`)

Archcry replied:
Well they're both optional, if you don't specify any of them it will load the config in your package.json. could clarify that.

nicojs replied:
I know it is optional, but if it has only one option, i would assume it is also the default. So specifying there are 2 options will help.

nicojs commented:
I know it is optional, but if it has only one option, i would assume it is
also the default. So specifying there are 2 options will help.

On Wed, Jan 31, 2018 at 2:51 PM, Sander Koenders <notifications@github.com>
wrote:

> *@Archcry* commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/stryker-mutator/stryker-jest-runner/pull/25#discussion_r165056845>
> :
>
> >      coverageAnalysis: 'off'
>  }
>  ```
>
> -### Loading the plugin
> +Stryker-jest-runner provides a couple of configurable options using the `jest` property in your stryker config:
> +
> +```javascript
> +{
> +    jest: {
> +        project: 'react',
> +        config: require('path/to/your/custom/jestConfig.js')
> +    }
> +}
> +```
> +
> +| option | value |
> +|----|----|
> +| project (optional) | The project you are working on (we currently only support "react")|
>
> Well they're both optional, if you don't specify any of them it will load
> the config in your package.json. could clarify that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stryker-mutator/stryker-jest-runner/pull/25#discussion_r165056845>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvliRm7VpUOW0hzGwj5V3qWiPGElaxpks5tQG_ZgaJpZM4RWIJw>
> .
>

nicojs commented:
@mthmulders can you live with this as well?

mthmulders commented on src/configLoaders/ReactConfigLoader.ts:
```diff
@@ -0,0 +1,35 @@
+import * as path from 'path';
+import JestConfiguration from './JestConfiguration';
+
+export default class ReactConfigLoader implements ConfigLoader {
```
I'm not so happy with this file / class name. It suggests you need this when you use React, which is not true. You need it when you use [Create React App (CRA)](https://github.com/facebookincubator/create-react-app). Maybe we should name it `ReactScriptsLoader` (because CRA uses `react-scripts` under the hood).

Archcry replied:
Well it isn't a loader for React scripts either. What about `ReactJestConfigLoader`, maybe `ConfigLoader` has to be renamed to `JestConfigLoader` and `DefaultConfigLoader` to `DefaultJestConfigLoader`. That way we clarify that it is loading a Jest configuration.

mthmulders commented:
Awesome work, @Archcry! It looks impressive and I'm overall happy with the results. There's two things I'd like to see addressed. These are not implementation issues but mainly about naming / bookkeeping.
